src/metabase/pulse/render/body.clj

```diff
@@ -642,41 +672,95 @@
 
 (def ^:private axis-group-threshold 0.33)
 
+(defn- group-axes-at-once
+  [joined-rows viz-settings threshold]
+  (let [;; a double-x-axis 'joined-row' looks like:
+        ;; [["val on x-axis"         "grouping-key"] [series-val]] eg:
+        ;; [["2016-01-01T00:00:00Z"  "Doohickey"   ] [9031.5578 ]]
+
+        ;; a single-x-axis 'joined-row' looks like:
+        ;; [[grouping-key] [series-val-1 series-val-2 ...]]
+        joined-rows-map    (if (= (count (ffirst joined-rows)) 2)
+                             ;; double-x-axis
+                             (-> (group-by (fn [[[_ x2] _]] x2) joined-rows)
+                                 (update-vals #(mapcat last %)))
+                             ;; single-x-axis
+                             (->> (:graph.metrics viz-settings)
+                                  (map-indexed (fn [idx k]
+                                                 [k (mapv #(get (second %) idx) joined-rows)]))
+                                  (into {})))
+        ;; map of group-key -> :left :right or nil
+        starting-positions (into {} (for [k (keys joined-rows-map)]
+                                      [k (keyword (series-setting viz-settings k :axis))]))
+        ;; map of position (:left :right or nil) -> vector of 'assigned' groups
+        positions          (-> (group-by second starting-positions)
+                               (update-vals #(mapv first %)))
+        check              {:unassigned? (contains? positions nil)
+                            :stacked?    (boolean (:stackable.stack_type viz-settings))} ]
+    (case check
+      {:unassigned? true :stacked? true} (into {} (map (fn [k] [k :left]) (keys joined-rows-map)))
+      {:unassigned? false :stacked? true} (into {} (map (fn [k] [k :left]) (keys joined-rows-map)))
+      {:unassigned? true :stacked? false}
+      (let [;; calculate the overlaps between every group.
+            ;; If there are 4 groups, there will be 4x4 overlaps calculated.
+            ;; TODO: the reciprocal pairings are technically unneeded calculation. It's optimizable,
+            ;; but the number of total groupings might never be high enough to really worry about it.
+            overlaps           (-> joined-rows-map
+                                   (update-vals (fn [vs]
+                                                  (into {} (map (fn [k]
+                                                                  [k (overlap (get joined-rows-map k) vs)])
+                                                                (keys joined-rows-map))))))
```

Comment:
Looks like the only use of `overlaps` is to get `relevant-distances`. Can't you just compute `relevant-distances` directly and skip creating the entire 2-d map of overlaps?

Reply:
Yea, that's smart. Haha, thanks for pointing it out. I do that with `score-fn` instead now. Simpler!